test/xds/xds_client_ignore_resource_deletion_test.go

```diff
@@ -0,0 +1,474 @@
+	if err != nil {
+		t.Fatalf("failed to dial local test server: %v", err)
+	}
+	defer cc.Close()
+
+	verifyRPCtoAllEndpoints(t, cc)
+
+	// Update without a listener resource.
+	if err := ms.Update(ctx, e2e.UpdateOptions{
+		NodeID:    nodeID,
+		Listeners: []*listenerpb.Listener{},
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	// Make RPCs every 100 ms for 1s and verify that the serving mode does
+	// not change on server.
+	timer := time.NewTimer(500 * time.Millisecond)
+	ticker := time.NewTicker(50 * time.Millisecond)
+	t.Cleanup(ticker.Stop)
+	for {
+		verifyRPCtoAllEndpoints(t, cc)
+		select {
+		case <-timer.C:
+			return
+		case mode := <-updateCh:
+			t.Fatalf("Listener received new mode: %v", mode)
+		case <-ticker.C:
+		}
+	}
+}
+
+// This test creates a xds-enabled gRPC server with a listener resource and this
+// server does not feature "ignore_resource_deletion". The test verifies RPCs are
+// working as expected. The test then removes the listener resource from the server
+// and verifies that server should enter "non_serving" mode as expected.
+func (s) TestListenerResourceDeletionOnServerNotIgnored(t *testing.T) {
+	ms := startManagementServer(t)
+	bootstrapContents := generateBootstrapContents(t, ms.Address, false)
+	resources, lis := resourceWithListenerForGRPCServer(t)
+	updateCh, serve := setupGRPCServerWithModeChangeChannel(t, bootstrapContents, lis)
+
+	go serve()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := ms.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Wait for the listener to move to "serving" mode.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Test timed out waiting for a mode change update.")
+	case mode := <-updateCh:
+		if mode != connectivity.ServingModeServing {
+			t.Fatalf("Listener received new mode %v, want %v", mode, connectivity.ServingModeServing)
+		}
+	}
+
+	// Create a ClientConn and make a successful RPCs.
+	cResolver := generateCustomResolver(t, bootstrapContents)
+	cc, err := grpc.Dial(lis.Addr().String(), grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithResolvers(cResolver))
+	if err != nil {
+		t.Fatalf("failed to dial local test server: %v", err)
+	}
+	defer cc.Close()
+	verifyRPCtoAllEndpoints(t, cc)
+
+	if err := ms.Update(ctx, e2e.UpdateOptions{
+		NodeID:    nodeID,
+		Listeners: []*listenerpb.Listener{}, // empty listener resource
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timed out waiting for a mode change update: %v", err)
+	case mode := <-updateCh:
+		if mode != connectivity.ServingModeNotServing {
+			t.Fatalf("listener received new mode %v, want %v", mode, connectivity.ServingModeNotServing)
+		}
+	}
+}
+
+// This helper makes both UnaryCall and EmptyCall RPCs using the ClientConn that
+// is passed to this function. This helper panics for any failed RPCs.
+func verifyRPCtoAllEndpoints(t *testing.T, cc grpc.ClientConnInterface) {
+	t.Helper()
```

Comment:
Back to the helper topic. This function is actually performing some test logic and would be better to not make it a t.Helper and instead return the error to the test. Doing so *will* make the actual test a tiny bit longer, but that should be OK.